src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4620 @@
+        num_attention_heads,  # Split channels across attention heads
+        num_channels // num_attention_heads,  # Channels per head
+    )
+
+    # Permute the dimensions to bring attention heads next to the spatial patches, keeping timesteps and the per-head channels intact.
+    windows = input_feature.permute(0, 2, 4, 6, 1, 3, 5, 7).contiguous()
+    return windows
+
+
+class ProPainterSparseWindowAttention(nn.Module):
+    def __init__(
+        self,
+        hidden_size: int,
+        num_attention_heads: int,
+        window_size: Tuple[int, int],
+        pool_size: Tuple[int, int] = (4, 4),
+        qkv_bias: bool = True,
+        attn_drop: float = 0.0,
+        proj_drop: float = 0.0,
+        pooling_token: bool = True,
+    ):
+        super().__init__()
+        assert hidden_size % num_attention_heads == 0
+        # key, query, value projections for all heads
+        self.key = nn.Linear(hidden_size, hidden_size, qkv_bias)
+        self.query = nn.Linear(hidden_size, hidden_size, qkv_bias)
+        self.value = nn.Linear(hidden_size, hidden_size, qkv_bias)
+        # regularization
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.proj_drop = nn.Dropout(proj_drop)
+        # output projection
+        self.proj = nn.Linear(hidden_size, hidden_size)
+        self.num_attention_heads = num_attention_heads
+        self.window_size = window_size
+        self.pooling_token = pooling_token
+        if self.pooling_token:
+            kernel_size, stride = pool_size, pool_size
+            self.pool_layer = nn.Conv2d(
+                hidden_size,
+                hidden_size,
+                kernel_size=kernel_size,
+                stride=stride,
+                padding=(0, 0),
+                groups=hidden_size,
+            )
+            self.pool_layer.weight.data.fill_(1.0 / (pool_size[0] * pool_size[1]))
+            self.pool_layer.bias.data.fill_(0)
+        self.expand_size = tuple((i + 1) // 2 for i in window_size)
+
+        if any(i > 0 for i in self.expand_size):
+            # get mask for rolled k and rolled v
+            mask_tl = torch.ones(self.window_size[0], self.window_size[1])
+            mask_tl[: -self.expand_size[0], : -self.expand_size[1]] = 0
+            mask_tr = torch.ones(self.window_size[0], self.window_size[1])
+            mask_tr[: -self.expand_size[0], self.expand_size[1] :] = 0
+            mask_bl = torch.ones(self.window_size[0], self.window_size[1])
+            mask_bl[self.expand_size[0] :, : -self.expand_size[1]] = 0
+            mask_br = torch.ones(self.window_size[0], self.window_size[1])
+            mask_br[self.expand_size[0] :, self.expand_size[1] :] = 0
+            masked_rolled_key = torch.stack((mask_tl, mask_tr, mask_bl, mask_br), 0).flatten(0)
+            self.register_buffer("valid_ind_rolled", masked_rolled_key.nonzero(as_tuple=False).view(-1))
+
+        self.max_pool = nn.MaxPool2d(window_size, window_size, (0, 0))
+
+    def forward(
```

Comment:
considering the mighty size of this forward it's acceptable to cut it down into 2 sub-methods at least, for each branching